src/workerd/api/streams/readable-source-adapter.c++

```diff
@@ -612,7 +613,125 @@ ReadableStreamSourceKjAdapter::~ReadableStreamSourceKjAdapter() noexcept(false)
+      KJ_IF_SOME(leftOverData, leftOver) {
+        auto ptr = leftOverData.asPtr();
+        return Active::Readable{
+          .data = kj::mv(leftOverData),
+          .view = ptr,
+        };
+      }
+      return kj::none;
+    }
+  };
+
+  // Copies as much data from source into the context as possible, returning
+  // the number of bytes copied.
+  static const auto copyFromSource = [](jsg::Lock& js, ReadContext& context,
+                                         const JsByteSource& source) -> CopyResult {
+    KJ_SWITCH_ONEOF(source) {
+      KJ_CASE_ONEOF(str, jsg::JsRef<jsg::JsString>) {
+        auto view = str.getHandle(js);
+        size_t len = view.length(js);
+        size_t toCopy = kj::min(len, context.buffer.size());
+
+        if (toCopy == 0) {
+          // Copy nothing. Return 0.
+          return {0};
+        }
+
+        if (toCopy < len) {
+          // We are going to have left-over data. Unfortunately in this case
+          // we have to copy the data twice... once into a kj::String and
+          // again into our buffer. This is because the V8 string UTF-8
+          // write API does not support partial writes with an offset.
+          auto data = view.toUSVString(js);
+          context.buffer.first(toCopy).copyFrom(data.asBytes().first(toCopy));
+          context.totalRead += toCopy;
+          context.buffer = context.buffer.slice(toCopy);
+          KJ_DASSERT(context.buffer.size() == 0);
+          return {toCopy, kj::Maybe(data.asBytes().slice(toCopy).attach(kj::mv(data)))};
+        }
+
+        // We can copy everything in one go. Yay! This is great because we
+        // can avoid a double copy here.
+        auto ret = view.writeInto(js, context.buffer.asChars().first(toCopy),
+            jsg::JsString::WriteFlags::REPLACE_INVALID_UTF8);
+        KJ_DASSERT(ret.written == toCopy);
+        context.totalRead += toCopy;
+        context.buffer = context.buffer.slice(toCopy);
+        return {toCopy};
+      }
+      KJ_CASE_ONEOF(ab, jsg::JsRef<jsg::JsArrayBuffer>) {
+        auto src = ab.getHandle(js).asArrayPtr();
+        size_t toCopy = kj::min(src.size(), context.buffer.size());
+        if (toCopy == 0) {
+          // Copy nothing. Return 0.
+          return {0};
+        }
+
+        context.buffer.first(toCopy).copyFrom(src.first(toCopy));
+        context.totalRead += toCopy;
+        context.buffer = context.buffer.slice(toCopy);
+
+        if (toCopy < src.size()) {
+          KJ_DASSERT(context.buffer.size() == 0);
+          // For now, we have to copy the left-over data into a new array.
+          // Why? I'm happy you asked! Because the src is backed by a
+          // v8::BackingStore protected by the v8 sandboxing rules and we
+          // don't yet have the memory protection key logic in place to safely
+          // share that memory outside of the v8 heap. For now, copy. Later
+          // we can revisit this to hopefully avoid the additinal copy.
```

Comment:
Could leave a `TODO(mpk)` here and make a note in some MPK-related ticket to grep for `TODO(mpk)` later.